Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+        const double *y = query.m_coords[1];
+#if AMREX_SPACEDIM == 3
+        const double *z = query.m_coords[2];
+#endif
+
+        // Run a check on coords you are interpolating on
+        for (int i = 0; i < n; i++)
+        {
+            std::array<double, AMREX_SPACEDIM> coords{x[i], y[i]
+#if AMREX_SPACEDIM == 3
+                                                      ,
+                                                      z[i]
+#endif
+            };
+            check_domain(coords, 0);
+        }
+
+        // copy stuff as otherwise I get errors with lambdas when I use class
+        // members inside the GPU loops; this is a bit annoying as I have to
+        // copy a few things; any other alternative ways?
+        const auto prob_lo    = m_prob_lo;
+        const auto prob_hi    = m_prob_hi;
+        const auto lo_reflect = m_lo_boundary_reflective;
+        const auto hi_reflect = m_hi_boundary_reflective;
+        const int ncomp       = num_components;
+
+        // copy coords here
+        const amrex::Real *x_p = nullptr, *y_p = nullptr, *z_p = nullptr;
+        amrex::Gpu::DeviceVector<amrex::Real> Xd, Yd, Zd;
+
+        // copy x,y,z to device vectors now
+        amrex::Gpu::DeviceVector<double> X(n), Y(n);
+        amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice, x, x + n, X.begin());
+        amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice, y, y + n, Y.begin());
+
+#if AMREX_SPACEDIM == 3
+        amrex::Gpu::DeviceVector<double> Z(n);
+        amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice, z, z + n, Z.begin());
+#endif
+        amrex::Gpu::streamSynchronize(); // ensure copies complete
+
+        // Get raw device pointers to capture by value
+        auto x_d = X.data();
+        auto y_d = Y.data();
+#if AMREX_SPACEDIM == 3
+        auto z_d = Z.data();
+#endif
+
+        // loop over particles and place them at the required points
+        amrex::ParallelFor(
+            n,
+            [=] AMREX_GPU_DEVICE(int ip)
+            {
+                auto &p = particle_data[ip];
+                p.id()  = ip + 1; // particle id starts from 1
+                p.cpu() = 0;      // CPU id, not used here
+
+                // reflect into valid region for seeding
+                const amrex::Real xr = reflect_particle(
+                    static_cast<amrex::Real>(x_d[ip]), prob_lo[0], prob_hi[0],
+                    lo_reflect[0], hi_reflect[0]);
+                const amrex::Real yr = reflect_particle(
+                    static_cast<amrex::Real>(y_d[ip]), prob_lo[1], prob_hi[1],
+                    lo_reflect[1], hi_reflect[1]);
+#if AMREX_SPACEDIM == 3
+                const amrex::Real zr = reflect_particle(
+                    static_cast<amrex::Real>(z_d[ip]), prob_lo[2], prob_hi[2],
+                    lo_reflect[2], hi_reflect[2]);
+#endif
+
+                // set position
+                p.pos(0) = xr;
+                p.pos(1) = yr;
+#if AMREX_SPACEDIM == 3
+                p.pos(2) = zr;
+#endif
```

Comment:
Might as well simplify this.
```suggestion
                // reflect into valid region and set
                p.pos(0) = reflect_particle(
                    static_cast<amrex::Real>(x_d[ip]), prob_lo[0], prob_hi[0],
                    lo_reflect[0], hi_reflect[0]);
                p.pos(1) = reflect_particle(
                    static_cast<amrex::Real>(y_d[ip]), prob_lo[1], prob_hi[1],
                    lo_reflect[1], hi_reflect[1]);
#if AMREX_SPACEDIM == 3
                p.pos(2) = reflect_particle(
                    static_cast<amrex::Real>(z_d[ip]), prob_lo[2], prob_hi[2],
                    lo_reflect[2], hi_reflect[2]);
#endif
```